### PR TITLE
Read excel nrows

### LINKED ISSUE #16645: Pandas read_excel: only read first few lines
#### Code Sample, a copy-pastable example if possible

```python
workbook_dataframe = pd.read_excel(workbook_filename, nrows = 10)
```
#### Problem description

Using pandas read_excel on about 100 excel files - some are large - I want to read the first few lines of each (header and first few rows of data).

The above doesn't work but illustrates the goal (example reading 10 data rows).

### PR DESCRIPTION
**Pull request by alysivji**
- [x] closes #16645
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -38,6 +38,7 @@ Other Enhancements
 - :func:`read_feather` has gained the ``nthreads`` parameter for multi-threaded operations (:issue:`16359`)
 - :func:`DataFrame.clip()` and :func: `Series.cip()` have gained an inplace argument. (:issue: `15388`)
 - :func:`crosstab` has gained a ``margins_name`` parameter to define the name of the row / column that will contain the totals when margins=True. (:issue:`15972`)
+- ``pd.read_excel()`` has a ``nrows`` parameter (:issue:`16645`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.21.0.txt]: How about:

`pd.read_excel()` has gained the `nrows` parameter (...)

File: pandas/io/excel.py [hunk 2]
```diff
@@ -191,12 +193,12 @@ def get_writer(engine_name):
...
 
 @Appender(_read_excel_doc)
-def read_excel(io, sheet_name=0, header=0, skiprows=None, skip_footer=0,
-               index_col=None, names=None, parse_cols=None, parse_dates=False,
-               date_parser=None, na_values=None, thousands=None,
-               convert_float=True, converters=None, dtype=None,
-               true_values=None, false_values=None, engine=None,
-               squeeze=False, **kwds):
+def read_excel(io, sheet_name=0, header=0, skiprows=None, nrows=None,
+               skip_footer=0, index_col=None, names=None, parse_cols=None,
```
Comment (jreback) [inline pandas/io/excel.py]: we normally don't like to shuffle parameters around in kwargs. Please align the ordering of these params as much as possible with how read_csv does it (obviously only include the current parameters).

File: pandas/tests/io/test_excel.py [hunk 3]
```diff
@@ -1000,6 +1000,20 @@ def test_read_excel_skiprows_list(self):
...
+        # GH 16645
+        num_rows_to_pull = 5
+        actual = pd.read_excel(os.path.join(self.dirpath, 'test1' + self.ext),
+                               nrows=num_rows_to_pull)
+        expected = pd.read_excel(os.path.join(self.dirpath,
+                                              'test1' + self.ext))
+        expected = expected[:num_rows_to_pull]
+        tm.assert_frame_equal(actual, expected)
+
+        with pytest.raises(ValueError):
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: @gfyoung don't we have an issue about ``nrows`` validation in the parser?

File: pandas/tests/io/test_excel.py [hunk 4]
```diff
@@ -1000,6 +1000,20 @@ def test_read_excel_skiprows_list(self):
...
 
+    def test_read_excel_nrows(self):
+        # GH 16645
+        num_rows_to_pull = 5
+        actual = pd.read_excel(os.path.join(self.dirpath, 'test1' + self.ext),
+                               nrows=num_rows_to_pull)
+        expected = pd.read_excel(os.path.join(self.dirpath,
+                                              'test1' + self.ext))
+        expected = expected[:num_rows_to_pull]
+        tm.assert_frame_equal(actual, expected)
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: try to pull *more* rows than exist in the file as well.

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -999,6 +999,8 @@ def _failover_to_python(self):
 
     def read(self, nrows=None):
         if nrows is not None:
+            nrows = _validate_integer('nrows', nrows)
```
Comment (gfyoung) [inline pandas/io/parsers.py]: I think this is unnecessary.  This already gets validated the moment you call `read_excel`.

File: pandas/tests/io/test_excel.py (hunk 3 above)
Comment (gfyoung) [inline pandas/tests/io/test_excel.py]: 1) We have an entire test suite for this in `common.py` (under `io/parsers`)

2) This already gets validated.  Thus, this test is somewhat unnecessary.

Comment (gfyoung) [inline pandas/tests/io/test_excel.py]: @alysivji : Put this as a separate test, and use `tm.assert_raises_regex` to check the specific error message raised in the `ValueError` (we want it to be that validation failed).

File: pandas/io/parsers.py (hunk 5 above)
Comment (alysivji) [inline pandas/io/parsers.py]: I originally didn't have this line there, but then I was getting the following error:
```TypeError: '>=' not supported between instances of 'int' and 'str'``` vs
```ValueError: 'nrows' must be an integer >=0```

File: pandas/io/excel.py (hunk 2 above)
Comment (alysivji) [inline pandas/io/excel.py]: Sure, not a problem. I was using read_csv as a guide and saw that ```nrows``` was directly after ```skiprows```, hence the change.

I will go ahead and line up the read_excel kwargs as close as possible to the read_csv kwargs. I see 3or4 out of order with just a quick glance.

Comment (jreback) [inline pandas/io/excel.py]: ok great. prob need a slightly expanded whatsnew note to tell about this

Comment (alysivji) [inline pandas/io/excel.py]: I rearranged the kwargs and also updated docstrings to match parameter order (one of the things that always bugs me). Should I also update all internal function kwargs in ```excel.py``` to match the external API?

I'm adding the following note to the Other Enhancements section (just wanted to make sure it was the right spot...):
```- Rearranged the order of keyword arguments in :func:`read_excel()` to align with :func:`read_csv()` ```

Thanks!

File: pandas/io/excel.py [hunk 6]
```diff
@@ -82,6 +82,8 @@
     Rows to skip at the beginning (0-indexed)
 skip_footer : int, default 0
     Rows at the end to skip (0-indexed)
+nrows : int, default None
+    Number of rows to parse
```
Comment (jreback) [inline pandas/io/excel.py]: versionadded 0.21.0 tag

File: pandas/io/excel.py (hunk 2 above)
Comment (jreback) [inline pandas/io/excel.py]: yes I would have the internal API match the external

File: doc/source/whatsnew/v0.21.0.txt (hunk 1 above)
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: you can make another entry (with this PR number), in api_breaking section that the kwargs are re-aranged to match pd.read_csv

use ``:func:`read_excel` ``

File: pandas/io/parsers.py (hunk 5 above)
Comment (jreback) [inline pandas/io/parsers.py]: yes this change should be done in the parser itself. See if you can come up with an example that ONLY used pd.read_csv directly.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 5 above)
Comment (alysivji) [inline pandas/io/parsers.py]: Where should I put this test? There are a bunch in `tests/io/parser', but nothing for read_csv directly.

**Comment by jreback** [comment]
can you rebase/update

**Comment by jreback** [comment]
can you rebase / update

**Comment by alysivji** [comment]
Sure, I got some time this week. Will take a look at it.

**Comment by jreback** [comment]
pls rebase

**Comment by jreback** [comment]
closing as stale. if you'd like to continue, pls ping.

**Comment by alysivji** [comment]
I got some time. Picked this back up. Sent another PR (18507)

Thanks.

**Pull request closed without merging**